src/frog-base.tsx

```diff
@@ -385,12 +375,68 @@ export class FrogBase<
           return defaultImageOptions?.fonts
         })()
 
-        const {
-          headers = this.headers,
-          image,
-          imageOptions = defaultImageOptions,
-        } = fromQuery<any>(query)
-        const image_ = JSON.parse(lz.decompressFromEncodedURIComponent(image))
+        const image_ = image
+          ? JSON.parse(lz.decompressFromEncodedURIComponent(image))
+          : await (async () => {
+              // Request body will be empty since a request for the image is always done with
+              // `GET` http method. Thus, the result of the next invocation is intristically the same
+              // as getting frame context for an initial frame.
+              const { context } = getFrameContext<env, string>({
+                context: await requestBodyToContext(c, {
+                  hub:
+                    this.hub ||
+                    (this.hubApiUrl ? { apiUrl: this.hubApiUrl } : undefined),
+                  secret: this.secret,
+                  verify,
+                }),
+                initialState: this._initialState,
+                origin,
+              })
+              const response = await handler(context)
```

Comment:
If we go down this route, then this will cause a double render cycle on the initial request (one render cycle on the main endpoint, and one on this `/image` endpoint). This means it will become non-idempotent (ie. operations in the handler will be invoked twice which could cause db/store related bugs).

I wonder if there is a way to do this without invoking the handler again.

Reply:
I still need to play around with the initial image caching issue, but why would WC cache the initial image if `cache-control=0` on the frame endpoint? Shouldn’t it always serve fresh frame data and seems like a WC issue?

Reply:
The issue is that WC caches an image by url, and if an image changes its url also changes because the compressed `image` query parameter is changing.

Reply:
Therefore since every "refreshed" image has a unique link, wc will cache the initial one and would expect the initial one to be updated.

Reply:
> If we go down this route, then this will cause a double render cycle on the initial request (one render cycle on the main endpoint, and one on this `/image` endpoint). This means it will become non-idempotent (ie. operations in the handler will be invoked twice which could cause db/store related bugs).
> 
> 
> 
> I wonder if there is a way to do this without invoking the handler again.

You're absolutely right.
We need to think of a different solution then.

Reply:
But just to note – the idempotency would be broken only for the initial frame.

Reply:
> I still need to play around with the initial image caching issue, but why would WC cache the initial image if `cache-control: max-age=0` on the frame endpoint? Shouldn’t it always serve fresh frame data and seems like a WC issue?

Also another thing to note is that `cache-control` header has to be set not in the frame response but in the frame image response.

Reply:
Yeah, I guess what I am trying to say is that, should WC also respect headers in the frame response? Seems like it should.

Reply:
+, this would solve many issues

Reply:
That's why I said that we should first reach out to the wc team. I have a feeling that they're not handling that the way it's expected.

Reply:
https://warpcast.com/horsefacts.eth/0x66aac3d2 IMO the cleaner way to handle this. It's a small step for the user but the code is even cleaner.

Reply:
@jxom, if idempotency is our main issue in this PR, we could enforce frame image handler to **always** be invoked after the frame handler.

Currently, frog implements those to be independent from each other - meaning that given a frame response you're given a specific frame image URL that always renders "independently" from a frame. As in fact if you serialize the query parameters correctly, you could render any image even outside of the frame context.

However, all the app clients are fetching frames first, and retrieve image urls as a result, thus the frame image in fact depends on the frame response itself.

I propose to implement an initial frame image storage inside of frog which would enforce triggering only one of the handlers and returning a stores value in another one.

Here's the first example, where a Frame is casted for the first time (meaning that frame is not cached in the app client):
1. App client requests a frame. Frame handler is executed, where image is one of the returns. If the `context.status === "initial"`, we put the `image` in a local storage where the key is some kind of unique `requestId` (can we get it from somewhere?), which stays the same within a single full frame fetch operation..
2. App client requests for frame image. Image handler is executed. Since frame handler was executed before, we have the `image` in local storage by the given `requestId`, so we just return it.

And here's the second example where the app client has the frame cached.
The example is also only applicable for requests where `context.status === "initial"`
1. Since the frame response is cached **indefinitely**, app client makes a fetch for frame image (given that the `age` of the cached frame image request has surpassed `max-age`).
2. Local storage doesn't have a cached value for that specific `requestId` (because frame handler was not invoked), so the image handler invokes frame handler just to retrieve the image, and doesn't store anything in the local storage since there will be no requests afterwards.

Reply:
I've seen more and more errors pop up. Until a decision is made can we rollback to where it was before the "fetch" from within function was implemented? It's getting critical IMO.

Reply:
> @jxom, if idempotency is our main issue in this PR, we could enforce frame image handler to **always** be invoked after the frame handler.
> 
> Currently, frog implements those to be independent from each other - meaning that given a frame response you're given a specific frame image URL that always renders "independently" from a frame. As in fact if you serialize the query parameters correctly, you could render any image even outside of the frame context.
> 
> However, all the app clients are fetching frames first, and retrieve image urls as a result, thus the frame image in fact depends on the frame response itself.
> 
> I propose to implement an initial frame image storage inside of frog which would enforce triggering only one of the handlers and returning a stores value in another one.
> 
> Here's the first example, where a Frame is casted for the first time (meaning that frame is not cached in the app client):
> 
> 1. App client requests a frame. Frame handler is executed, where image is one of the returns. If the `context.status === "initial"`, we put the `image` in a local storage where the key is some kind of unique `requestId` (can we get it from somewhere?), which stays the same within a single full frame fetch operation..
> 2. App client requests for frame image. Image handler is executed. Since frame handler was executed before, we have the `image` in local storage by the given `requestId`, so we just return it.
> 
> And here's the second example where the app client has the frame cached. The example is also only applicable for requests where `context.status === "initial"`
> 
> 1. Since the frame response is cached **indefinitely**, app client makes a fetch for frame image (given that the `age` of the cached frame image request has surpassed `max-age`).
> 2. Local storage doesn't have a cached value for that specific `requestId` (because frame handler was not invoked), so the image handler invokes frame handler just to retrieve the image, and doesn't store anything in the local storage since there will be no requests afterwards.

just figured out that all of that is applicable without this `requestId`, there's no need to preserve the "origin" of the request as the context is empty – GET requests have no payload in them.

Reply:
such a "storage" most likely won't be reliable in serverless environments where a worker would kill itself in between frame and frame image requests causing idempotency breakage

Reply:
> such a "storage" most likely won't be reliable in serverless environments where a worker would kill itself in between frame and frame image requests causing idempotency breakage

however, the delay between such request is so small that it's less likely that a pod would go to sleep